py/desisim/archetypes.py

```diff
@@ -29,29 +29,42 @@ def compute_chi2(flux, ferr=None):
     -------
     Tuple of (chi2, amp) where:
         chi2 : numpy.ndarray
-            Chi^2 matrix [Nspec, Nspec] between all combinations of spectra.
+            Chi^2 matrix [Nspec, Nspec] between all combinations of normalized spectra.
         amp : numpy.ndarray
             Amplitude matrix [Nspec, Nspec] between all combinations of spectra.
 
     """
-    from SetCoverPy import mathutils    
 
     nspec, npix = flux.shape
+    chi2 = np.zeros((nspec, nspec), dtype='f4')
+    amp = np.zeros((nspec, nspec), dtype='f4')
+
+    flux = flux.copy()
+    rescale = np.sqrt(npix/np.sum(flux**2,axis=1))
+    flux *= rescale[:,None]
+
     if ferr is None:
-        ferr = np.ones_like(flux)
-
-    chi2 = np.zeros((nspec, nspec)).astype('f4')
-    amp = np.zeros((nspec, nspec)).astype('f4')
-
-    for ii in range(nspec):
-        if ii % 500 == 0 or ii == 0:
-            log.info('Computing chi2 matrix for spectra {}-{} out of {}.'.format(
-                (ii % 500) * 500, np.min(((ii+1) * (ii % 500), nspec-1)), nspec))
-        xx = flux[ii, :].reshape(1, npix)
-        xxerr = ferr[ii, :].reshape(1, npix)
-        amp1, chi21 = mathutils.quick_amplitude(xx, flux, xxerr, ferr, niter=1)
-        chi2[ii, :] = chi21
-        amp[ii, :] = amp1
+        for ii in range(nspec):
+            if ii % 500 == 0:
+                log.info('Computing chi2 matrix for spectra {}-{} out of {}.'.format(
+                    (ii % 500) * 500, np.min(((ii+1) * (ii % 500), nspec-1)), nspec))
+            amp1 = np.sum(flux[ii]*flux,axis=1)/npix
+            chi2[ii,:] = npix*(1.-amp1**2)
+            amp[ii,:] = amp1
+    else:
+        from SetCoverPy import mathutils
+        ferr *= rescale[:,None]
+        for ii in range(nspec):
+            if ii % 500 == 0 or ii == 0:
+                log.info('Computing chi2 matrix for spectra {}-{} out of {}.'.format(
+                    (ii % 500) * 500, np.min(((ii+1) * (ii % 500), nspec-1)), nspec))
+            xx = flux[ii, :].reshape(1, npix)
+            xxerr = ferr[ii, :].reshape(1, npix)
+            amp[ii,:], chi2[ii,:] = mathutils.quick_amplitude(xx, flux, xxerr, ferr, niter=1)
+
```

Comment:
In #357 you found that the chi^2 value returned by `SetCoverPy.mathutils.quick_amplitude` was wrong, so why are we using it at all?  

In case it's useful, here's another calculation of chi2 that takes into account the inverse variance, but I haven't tested how it compares to your algorithm or to `quick_amplitude`.

```
ivar = 1 / ferr**2
amplitude = np.sum(ivar[:, :, np.newaxis] * flux[:, :, np.newaxis] * \
  flux[:, np.newaxis, :], axis=0) ) / np.sum(ivar[:, :, np.newaxis] * \
  flux[:, np.newaxis, :]**2, axis=0)

chi2 = np.sum(ivar[:, :, np.newaxis] * ( flux[:, :, np.newaxis] - \
  amplitude[np.newaxis, :, :] * flux[:, np.newaxis, :])**2, axis=0)
```

Reply:
Actually it is correct once the spectra are normalized. So both piece of code give the same chi2 matrix.